### DISCUSSION TITLE
Data not saved into Google Cloud Storage

### DISCUSSION THREAD

**Discussion by dhanangw** [milvus-io/milvus#17710] [Q&A and General discussion] [answered]
I've deployed Miilvus cluster to GCP (Google Cloud Platform) in GKE (Google Kubernetes Engine) with Helm, and I use a new GCS (Google Cloud Storage) bucket for it's persistent storage.

I've followed Milvus' official deployment [guide](https://milvus.io/docs/gcp.md) for GCP during the deployment process,
and I ran Pymilvus' official [example](https://milvus.io/docs/v2.0.x/example_code.md) (with the part for deleting collection commented out) for verifying my deployment. 
The data is able to be searched but is not stored in GCS, as the GCS bucket is still empty.

In the FAQ section, it says the collection needs to be flushed to make it stored in the object storage.
I couldn't find the `flush()` method in Pymilvus 2.0.2.

Milvus has an `auto_flush_interval` [config](https://milvus.io/docs/v1.1.1/milvus_config.md#Section-storage) that defaults to 1s.
I didn't change the config, have waited for more than 1s, and my GCS bucket is still empty.

I have also made sure the IAM account used for MinIO GCS gateway has appropriate permissions.

Have i left something off? really need your guidance here.

Helm chart version: milvus-3.0.23
Milvus version: 2.0.2 
GKE cluster version: 1.21.11
Python version: 3.7.11
Pymilvus version: 2.0.2

**Comment by LoveEachDay** [comment, marked as answer]
Do you follow the https://milvus.io/docs/gcp.md example to create the access key and secret key? I suspect that iam account may not take effect in minio. And could you provide the minio log?

**Reply by dhanangw** [reply]
Hello @LoveEachDay 

Yes, I followed that GCP deployment guide.  
I use the default values for accesskey and secretkey.  
  
I found this error when checking minio pod 0:
```
API: SYSTEM()
Time: 02:26:44 UTC 06/13/2022
Error: Marking http://milvus-staging-minio-2.milvus-staging-minio-svc.milvus-staging.svc.cluster.local:9000/minio/storage/export/v43 temporary offline; caused by Post "http://milvus-staging-minio-2.milvus-staging-minio-svc.milvus-staging.svc.cluster.local:9000/minio/storage/export/v43/readall?disk-id=&file-path=format.json&volume=.minio.sys": lookup milvus-staging-minio-2.milvus-staging-minio-svc.milvus-staging.svc.cluster.local on 10.197.0.10:53: no such host (*fmt.wrapError)
       6: internal/rest/client.go:151:rest.(*Client).Call()
       5: cmd/storage-rest-client.go:152:cmd.(*storageRESTClient).call()
       4: cmd/storage-rest-client.go:520:cmd.(*storageRESTClient).ReadAll()
       3: cmd/format-erasure.go:387:cmd.loadFormatErasure()
       2: cmd/format-erasure.go:326:cmd.loadFormatErasureAll.func1()
       1: internal/sync/errgroup/errgroup.go:123:errgroup.(*Group).Go.func1()

API: SYSTEM()
Time: 02:26:44 UTC 06/13/2022
Error: Read failed. Insufficient number of disks online (*errors.errorString)
       5: cmd/prepare-storage.go:266:cmd.connectLoadInitFormats()
       4: cmd/prepare-storage.go:326:cmd.waitForFormatErasure()
       3: cmd/erasure-server-pool.go:91:cmd.newErasureServerPools()
       2: cmd/server-main.go:645:cmd.newObjectLayer()
       1: cmd/server-main.go:496:cmd.serverMain()
Waiting for a minimum of 2 disks to come online (elapsed 59s)

```
  
Followed by this warning:
```
WARNING: Detected default credentials 'minioadmin:minioadmin', we recommend that you change these values with 'MINIO_ROOT_USER' and 'MINIO_ROOT_PASSWORD' environment variables
Waiting for all MinIO sub-systems to be initialized.. lock acquired
Automatically configured API requests per node based on available memory on the system: 236
All MinIO sub-systems initialized successfully
Waiting for all MinIO IAM sub-system to be initialized.. lock acquired
Status:         4 Online, 0 Offline. 
API: http://10.84.20.229:9000  http://127.0.0.1:9000 

Console: http://10.84.20.229:44223 http://127.0.0.1:44223 

Documentation: https://docs.min.io

WARNING: Console endpoint is listening on a dynamic port (44223), please use --console-address ":PORT" to choose a static port.
```

**Reply by LoveEachDay** [reply]
The above error is transient. And `Status:         4 Online, 0 Offline.` shows that minio cluster is ready. 
 Could you provide a `kubectl describe <minio-pod>`?

**Reply by dhanangw** [reply]
This is the output of kubectl describe:
```
➜  milvus kubectl describe pod -n milvus-staging milvus-staging-minio-0
Name:         milvus-staging-minio-0
Namespace:    milvus-staging
Priority:     0
Node:         gke-ase1-glo-data-1-e2-std-4-sts-reg--40d91ac3-3qs2/172.16.20.55
Start Time:   Mon, 13 Jun 2022 09:25:24 +0700
Labels:       app=minio
              controller-revision-hash=milvus-staging-minio-7678f9d749
              release=milvus-staging
              statefulset.kubernetes.io/pod-name=milvus-staging-minio-0
Annotations:  checksum/config: 6d69d402b13dac9e68fae5f19b365e2821780a7d07e4dca755d4adc430bf1810
              checksum/secrets: 01ba4719c80b6fe911b091a7c05124b64eeece964e09c058ef8f9805daca546b
              cni.projectcalico.org/containerID: 8eac9c154c4f2513f9863dfb9d394b82362d0f101788991fa042a3ef0dace04a
              cni.projectcalico.org/podIP: 10.84.20.229/32
              cni.projectcalico.org/podIPs: 10.84.20.229/32
Status:       Running
IP:           10.84.20.229
IPs:
  IP:           10.84.20.229
Controlled By:  StatefulSet/milvus-staging-minio
Containers:
  minio:
    Container ID:  containerd://f025569210a7fde2d28dd7c3c14a314f2722d9032d3f2be6f803976fba57c73c
    Image:         minio/minio:RELEASE.2022-03-17T06-34-49Z
    Image ID:      docker.io/minio/minio@sha256:962f9543b7ce94b9df62944377c21718de0077a5ce3b18f7a723a18f34214d6f
    Port:          9000/TCP
    Host Port:     0/TCP
    Command:
      /bin/sh
      -ce
      /usr/bin/docker-entrypoint.sh minio -S /etc/minio/certs/ server  http://milvus-staging-minio-{0...3}.milvus-staging-minio-svc.milvus-staging.svc.cluster.local/export
    State:          Running
      Started:      Mon, 13 Jun 2022 09:25:43 +0700
    Ready:          True
    Restart Count:  0
    Requests:
      memory:  2Gi
    Environment:
      MINIO_ACCESS_KEY:  <set to the key 'accesskey' in secret 'milvus-staging-secret'>  Optional: false
      MINIO_SECRET_KEY:  <set to the key 'secretkey' in secret 'milvus-staging-secret'>  Optional: false
    Mounts:
      /export from export (rw)
      /var/run/secrets/kubernetes.io/serviceaccount from kube-api-access-47q2m (ro)
Conditions:
  Type              Status
  Initialized       True 
  Ready             True 
  ContainersReady   True 
  PodScheduled      True 
Volumes:
  export:
    Type:       PersistentVolumeClaim (a reference to a PersistentVolumeClaim in the same namespace)
    ClaimName:  export-milvus-staging-minio-0
    ReadOnly:   false
  minio-user:
    Type:        Secret (a volume populated by a Secret)
    SecretName:  milvus-staging-secret
    Optional:    false
  kube-api-access-47q2m:
    Type:                    Projected (a volume that contains injected data from multiple sources)
    TokenExpirationSeconds:  3607
    ConfigMapName:           kube-root-ca.crt
    ConfigMapOptional:       <nil>
    DownwardAPI:             true
QoS Class:                   Burstable
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:                      <none>

```

**Reply by LoveEachDay** [reply]
From the above output, your setting doesn't enabled gcsgateway. 
You can upgrade the milvus cluster with `--set minio.mode=standalone`

**Reply by LoveEachDay** [reply]
And when `helm upgrade` , you'd better with `--reuse-values --set minio.mode=standalone`, to upgrade milvus cluster while keeping previous settings.

**Reply by dhanangw** [reply]
I've included `--set minio.gcsgateway.enabled=true` param during the `helm install ...` part.  
and the variable is correctly enabled in the helm release
```
➜  milvus helm get values -n milvus-staging milvus-staging             
USER-SUPPLIED VALUES:
externalGcs:
  bucketName: rg-milvus
minio:
  existingSecret: milvus-staging-secret
  gcsgateway:
    enabled: true
    ...
...
...
```

**Reply by LoveEachDay** [reply]
You need both set `minio.gcsgateway.enabled=true` and `minio.mode=standalone`. 
You can find from the code file: `https://github.com/milvus-io/milvus-helm/blob/master/charts/minio/templates/deployment.yaml` where only `standalone` mode can enable `minio gateway` mode.

**Reply by dhanangw** [reply]
Hello @LoveEachDay 

Setting `minio.mode=standalone` solves the issue.  
Thank you very much for your help 🙇   
  
I've opened a [PR](https://github.com/milvus-io/milvus-docs/pull/1557) in the docs to add that missing value for `helm install ...` in GCP deployment guide page.

**Reply by juanmaro97** [reply]
Hello!

I could use GCS with the deployment of Milvus on GKE. 

I checked out the guide [here](https://milvus.io/docs/gcp.md#Use-Google-Cloud-Storage), when I run `helm install my-release milvus/milvus ...` it just does not work. The error is about "cannot reuse a name that is still in use", because the service of milvus has the same name. 

Then, I used `helm upgrade -i my-release milvus/milvus ....--reuse-values` and it did work 😊 

Milvus' docs are fine, but for the last command you should change `install` for `upgrade -i` and add `--reuse-values` argument too.

Do not forget `--set minio.mode=standalone` and `--set minio.gcsgateway.enabled=true`
